test/e2e/remove_orchestrator_test.go

```diff
@@ -0,0 +1,119 @@
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"net/url"
+	"testing"
+	"time"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/livepeer/go-livepeer/eth/contracts"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRemoveMOrchestrator(t *testing.T) {
+	// given
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	geth := setupGeth(t)
+	defer terminateGeth(t, geth)
+
+	// when
+	o := startAndRegisterOrchestrator(t, geth, ctx)
+	defer o.stop()
+
+	balance, _ := o.dev.Client.BalanceOf(o.dev.Client.Account().Address)
+
+	waitForNextRound(t, o.dev.Client)
+	o.dev.InitializeRound()
+
+	lockId := deactivateOrchestrator(o)
+
+	waitForNextRound(t, o.dev.Client)
+	o.dev.InitializeRound()
+
+	requireOrchestratorDeactivated(t, o)
+
+	lock, _ := o.dev.Client.GetDelegatorUnbondingLock(o.dev.Client.Account().Address, lockId)
+
+	waitUntilRound(lock.WithdrawRound, o, t)
+
+	withdrawStake(o, lockId)
+
+	// then
+	assertOrchestratorRemoved(t, o, balance)
+}
+
+func waitUntilRound(round *big.Int, o *livepeer, t *testing.T) {
+	for {
+		current, _ := o.dev.Client.CurrentRound()
+		if current.Cmp(round) == 0 {
+			return
+		}
+
+		waitForNextRound(t, o.dev.Client)
+		o.dev.InitializeRound()
+	}
+}
+
+func deactivateOrchestrator(o *livepeer) *big.Int {
+	val := url.Values{
+		"amount": {fmt.Sprintf("%d", lptStake)},
+	}
+
+	for {
+		txHash, ok := httpPostWithParams(fmt.Sprintf("http://%s/unbond", *o.cfg.CliAddr), val)
+		if ok {
+			receipt, err := o.dev.Client.Backend().TransactionReceipt(context.Background(), ethcommon.BytesToHash([]byte(txHash)))
+			if err == nil {
+				abi, err := contracts.BondingManagerMetaData.GetAbi()
+				if err == nil {
+					event := new(contracts.BondingManagerUnbond)
+					log := receipt.Logs[2]
+					abi.UnpackIntoInterface(event, "Unbond", log.Data)
+					return event.UnbondingLockId
+				}
+			}
+		}
+		time.Sleep(200 * time.Millisecond)
+	}
+}
+
+func assertOrchestratorRemoved(t *testing.T, o *livepeer, oldBalance *big.Int) {
+	require := require.New(t)
```

Comment:
I wonder about this check, shouldn't we just check that the orchestrator is not longer in the trancoder pool? That would mean that the orchestrator is removed. However, you check the balance here, I'm not sure I understand how it's related.

Reply:
There are two stages of o removal. By just unbonding (without withdrawal), the orchestrator is indeed no longer active and removed from the transcoder pool, but is still linked to the LP protocol by having his tokens still locked. For complete resignation, I think O needs to get its stake back to consider full removal.

Reply:
> There are two stages of o removal. By just unbonding (without withdrawal), the orchestrator is indeed no longer active and removed from the transcoder pool, but is still linked to the LP protocol by having his tokens still locked. For complete resignation, I think O needs to get its stake back to consider full removal.

Ok, I didn't know about it. I thought that removal just means that O disappears from the transcoder pool.

I'm not sure if we should test the "full removal" or not. I leave it up to you.



> It's true that the function body does not reflect its name. Beside the requirement of the PR
> 
> > Check that the orchestrator is unregistered and that stake is back in the ETH account
> 
> If we stick to `func assertOrchestratorRemoved` and want to assert O complete removal, then we should check that O is not present in the delegator's pool and that the unbonding lock entry is processed.
> 
> Checking balance should still be there but go to a separate assertion.
> 
> WDYT?

Yeah, in that case, I'd do one of the following:
- Rename it to `assertOrchestratorStakeIsZero()` or something similar OR
- Keep the name as it is, but explain in the comment why you check the balance to check if orchestrator is removed and also the meaning of the "full removal"